scaper/core.py

```diff
@@ -1913,29 +1986,94 @@ def _generate_audio(self, audio_path, ann, reverb=None,
                         'Unsupported event role: {:s}'.format(
                             e.value['role']))
 
+            # Finally combine all the files and optionally apply reverb.
+            # If there are no events, throw a warning.
+            if len(event_audio_list) == 0:
+                warnings.warn(
+                    "No events to synthesize (silent soundscape), no audio "
+                    "generated.", ScaperWarning)
+            else:                        
+                tfm = sox.Transformer()
+                if reverb is not None:
+                    tfm.reverb(reverberance=reverb * 100)
+
+                # Sum all events to get soundscape audio
+                soundscape_audio = sum(event_audio_list)
+
+                # Check for clipping and fix [optional]
+                max_sample = np.max(np.abs(soundscape_audio))
+                clipping = max_sample > 1
+                if clipping:
+                    warnings.warn('Soundscape audio is clipping!',
+                                  ScaperWarning)
+
+                if peak_normalization or (clipping and fix_clipping):
+
+                    # normalize soundscape audio and scale event audio
+                    soundscape_audio, event_audio_list, scale_factor = \
+                        peak_normalize(soundscape_audio, event_audio_list)
+
+                    ref_db_change = 20 * np.log10(scale_factor)
+
+                    if clipping and fix_clipping:
+                        warnings.warn(
+                            'Peak normalization applied to fix clipping with '
+                            'scale factor = {}. The actual ref_db of the '
+                            'generated soundscape audio will change by '
+                            'approximately {:.2f}dB with respect to the target '
+                            'ref_db of {})'.format(
+                                scale_factor, ref_db_change, self.ref_db),
+                            ScaperWarning)
+
+                    if scale_factor < 0.05:
+                        warnings.warn(
+                            'Scale factor for peak normalization is extreme '
+                            '(<0.05), event SNR values in the generated soundscape '
+                            'audio may not perfectly match their specified values.',
+                            ScaperWarning
+                        )
+
+                # Apply effects and reshape
+                soundscape_audio = tfm.build_array(
+                    input_array=soundscape_audio,
+                    sample_rate_in=self.sr,
+                )
+                soundscape_audio = soundscape_audio.reshape(-1, self.n_channels)
```

Comment:
So...just noticed that we do all of the clipping and peak normalization stuff BEFORE we apply reverb to the soundscape. Should these operations be reversed? Reverb could cause things to clip, or it could cause it to not be peak normalized anymore, etc. Sorry, I just noticed this was happening.

Reply:
I've re-ordered the operations. Current tests are passing. @pseeth is checking now whether applying reverb to a clipping signal and then normalization works as expected. Assuming it does, we'll stick to this ordering. We need to add unit tests to explicitly test for this scenario, I've added a comment on #134

Reply:
After looking into this, we shouldn't normalize. See #136.

Reply:
@pseeth has checked and the re-ordering fails. I've reverted to the original order: sum --> normalize --> reverb.